keras_spiking/model_energy.py

```diff
@@ -0,0 +1,810 @@
+
+        # --- print lines
+        line_strs = []
+        for i, line in enumerate(lines):
+            parts = []
+            for part, align, width in zip(line, col_aligns, col_widths):
+                fstr = "{:" + ("<" if i == 0 else align) + str(width) + "}"
+                part = fstr.format(part[:width])
+                parts.append(part)
+
+            line_strs.append("|".join(parts))
+
+            if i == 0:
+                line_strs.append("|".join("-" * width for width in col_widths))
+
+        print_fn = print if print_fn is None else print_fn
+        for line_str in line_strs:
+            print_fn(line_str)
+
+        # --- print totals
+        has_energy = False
+        has_spiking = False
+        for col, device in columns:
+
+            if col == "energy":
+                if not has_energy:
+                    print_fn("=" * len(line_strs[0]))
+                    has_energy = True
+
+                if self.devices[device]["spiking"]:
+                    has_spiking = True
+
+                energy = self.total_energy(
+                    device, dt=dt, timesteps_per_inference=timesteps_per_inference
+                )
+                print_fn(
+                    "Total energy per inference [Joules/inf] (%s): %0.2e"
+                    % (device, energy)
+                )
+        if print_warnings and has_energy:
+            print_fn(
+                textwrap.fill(
+                    "* These are estimates only; see the documentation for a list of "
+                    "the assumptions being made.",
+                    width=line_length,
+                    subsequent_indent="  ",
+                )
+            )
+        if (
+            print_warnings
+            and has_spiking
+            and not all(
+                self.layer_stats[layer]["spiking"] for layer in self.model.layers
+            )
+        ):
+            print_fn(
+                textwrap.fill(
+                    "* This model contains non-spiking activations that would not "
+                    "actually behave in the manner we assume in these calculations; "
+                    "we assume these layers will be converted to spiking equivalents. "
+                    "Consider using `keras_spiking.SpikingActivation` to make this "
+                    "conversion explicit.",
+                    width=line_length,
+                    subsequent_indent="  ",
+                )
+            )
+
+
+def _output_size(node):
+    output_shape = node.output_shapes[1:]
+    if any(x is None for x in output_shape):
+        raise ValueError(
+            "Cannot compute stats for '%s' layer because the output shape %s contains "
+            "undefined elements; if this represents a temporal axis, consider wrapping "
+            "this layer in ``tf.keras.layers.TimeDistributed``"
+            % (compat.node_layer(node).name, output_shape)
+        )
+    return np.prod(output_shape)
```

Comment:
So the place that this is going to break down is for layers that _can_ have time, but aren't wrapped by a `TimeDistributed` layer. For example, `Conv3D`, or even something like `Conv1D` if it's applied to time. That's why I took the initial approach of supplying the number of timesteps in the model in the `ModelEnergy` constructor. (For RNN layers, which users might want to implement custom stats for, I think it's safe to assume that there's always a time axis, the way we do for `SpikingActivation`. So this approach should work fine there.)

It's also a bit of a hassle for users to have to wrap e.g. Dense layers in a TimeDistributed wrapper, when for that layer type it makes no difference. Furthermore, in the case where the time axis is `None` then they'll get this nice error asking them to use TimeDistributed, but if they have a fixed number of timesteps, they wouldn't get this error, but the number of connections/neurons in that Dense layer would appear many times too large.

Reply:
The problem is that there's no guarantee that a `Conv1D/2D/3D` layer will have time as one of the axes. The previous implementation tried to guess that based on whether the shape is set to `None`, but that's equally problematic (there's plenty of reasons why you'd have `None` for the shape that don't involve time). So users have to wrap it in TimeDistributed to indicate the presence of time regardless (or some other, equivalent way of telling `ModelEnergy` that this layer should be considered temporal).

Reply:
No, in the previous implementation it was based on what you passed for `timesteps` to `ModelEnergy.__init__`. That's why I had all that structure for passing `timesteps` around. That basically indicated whether layers had a time axis or not.

You're right that if the second axis was `None`, then the previous implementation would require `timesteps=None`, which precludes having a `None` axis for things other than time. However, we're precluded from that in the current implementation, too, because we need to take `np.prod(output_shape)`, so if there's `None` in there, then we'll fail. So I don't see how we can support any of those other reasons for having `None` in the shape for reasons other than time.

Reply:
In the previous implementation, we could change it so rather than checking whether the second axis is `None`, and if so asserting that `timesteps=None`, we could do it the other way and if `timesteps=None` then assert that the second axis is `None`. This would mean that we're clearly not basing anything on whether the second axis is `None`. But as I said, it won't buy us anything, since we can't deal with `None` for things other than the batch size or the time axis.

Reply:
Yeah that is a decent point that we still can't support that functionality, so perhaps it's not worth guarding against. We'd have to use the static shapes inferred from `example_data`, but that's probably more work than we want to dive into in this PR. I'll think a little bit to see if there's a way to remove the `TimeDistributed` requirement without having to specify/pass around `timesteps`.

Reply:
Another factor to keep in mind is that the number of timesteps may not be constant throughout the model (as the `ModelEnergy(..., timesteps=x)` approach requires).  E.g., this model
```
inp = x = tf.keras.layers.Input(shape=(4, 10))
x = tf.keras.layers.Conv1D(filters=10, kernel_size=3)(x)
x = tf.keras.layers.Dense(units=10)(x)

model = tf.keras.Model(inp, x)

energy = keras_spiking.ModelEnergy(model, timesteps=4)
energy.summary()
```
incorrectly reports 50 synops for the dense layer (because it's assuming timesteps=4, while at that point the model actually has timesteps=2).

Reply:
I can't really think of a solution where you don't have to indicate which layers are temporal in some way or other. But we could have different ways of indicating which layers are temporal (like the `temporal_layers` approach I had before the squash commit), if that feels more streamlined. 

Another option is that we could add back the `None=time` assumption (since, as you say, we're not supporting other uses of None anyway). And then you'd only need the `TimeDistributed` wrapper for the `Time!=None` cases (which are rarer I'd say).

Reply:
So the previous code kind of accounted for this by dividing by the number of timesteps. So if you've got a first layer that runs every timestep (say we've got 16 timesteps), and a second layer that pools over time and thus runs for only 8 timesteps, with the previous code you'd divide all layers by 16 timesteps, so the connections/neurons in the second layer would be half of what they would be if it didn't pool over time. So while the neurons/connections appear to be less than the actual number of neurons/connections in the layer, these numbers accurately reflect how many neurons/connections are computed per timestep (since the layer is only running every second timestep). This would also work for layers that had no time, since they only have to run once, and we'd amortize that cost over all timesteps.

Reply:
As for a way to not have to pass `timesteps` around, we could just have a flag that indicates whether the model has time or not, and then in `_output_size` we just throw away the time axis if the model has time, and don't otherwise. The downsides to this are we lose that behaviour I just described for models where the number of timesteps changes throughout the model, and we'd also have to do something for the case where a model pools across time at the end of the model, and then has e.g. a dense layer with no time axis. (This is the problem I was trying to describe in scrum one day, and then I said that I'd just make some code to try to demonstrate it. Passing the timesteps around was my workaround for all this.)

Reply:
> So while the neurons/connections appear to be less than the actual number of neurons/connections in the layer, these numbers accurately reflect how many neurons/connections are computed per timestep (since the layer is only running every second timestep). 

I guess it depends how you think you'd map that model with variable timesteps on to the spiking device. I'd assume you end up having to run every layer every timestep, regardless of what the Keras model says. But it's true that that isn't super well defined.

Reply:
We've really got to assume that any given device can implement any given model. Some spiking devices might be able to handle variable timesteps, but you're right that Loihi definitely doesn't, and I'm not sure how easy it'd be on Spinnaker either.

I definitely think variable timesteps isn't something worth worrying a whole lot about; in the original implementation, it was kind of a side effect of trying to find a solution that worked for models that may have time in earlier layers and then pooled over time at the end. If we change the assumptions around variable timesteps (e.g. assume that each layer runs every timestep), that's fine with me.

Reply:
I pushed a fixup to show what I had in mind for the second idea above. So most of the time users will not need a TimeDistributed wrapper (which I think is the main thing you're concerned about?). 

So it works like:
```
inp = tf.keras.Input((None, 28))
x = tf.keras.layers.ReLU()(inp)
```
No TimeDistributed required, this is assumed to be a ReLU layer with 28 neurons operating over time

```
inp = tf.keras.Input((28, 28))
x = tf.keras.layers.ReLU()(inp)
```
No TimeDistributed required, this is assumed to be a ReLU layer with 784 neurons operating every timestep

```
inp = tf.keras.Input((28, 28))
x = tf.keras.layers.TimeDistributed(tf.keras.layers.ReLU())(inp)
```
Can use TimeDistributed to force ModelEnergy to interpret the layer temporally.

Reply:
So that works OK for `ReLU` (it basically forces you to use TimeDistributed only if it's a fixed number of timesteps, but not where `timesteps=None`), but it won't work for something like Conv3D, where the convolution is being applied over time, in the case of a fixed number of timesteps. I don't agree with the claim that a fixed number of timesteps is that uncommon; most of the networks that I've been using for gesture recognition have a fixed number of timesteps (since I know how long the videos are). That said, I think many networks with a fixed number of timesteps would also work for a variable number of timesteps, so we could basically say (in the docs): we don't support networks that have a fixed number of timesteps and don't use TimeDistributed for all layers with time.

Here's an example that worked correctly with the previous implementation, but doesn't with the current implementation
```
import numpy as np
import tensorflow as tf

import keras_spiking

timesteps = 10
# timesteps = None
image_shape = [3, 3, 4]

input_data = np.random.uniform(
    -1, 1, size=[2] + [10 if timesteps is None else timesteps] + image_shape
)

inp = tf.keras.Input([timesteps] + image_shape)
x = tf.keras.layers.Conv3D(
    filters=4, kernel_size=(3, 3, 3), padding="same", activation="relu"
)(inp)
model = tf.keras.Model(inp, [x])

if 0:
    # previous implementation
    energy = keras_spiking.ModelEnergy(
        model, input_data=input_data, timesteps=timesteps
    )
    energy.summary()
else:
    # current implementation
    energy = keras_spiking.ModelEnergy(
        model,
        example_data=input_data,
    )
    energy.summary()
```
It previously output
```
Layer (type)        |Output Shape         |Param #|Synops|Neurons
--------------------|---------------------|-------|------|-------
input_1 (InputLayer)|[(None, 10, 3, 3, 4)]|      0|     0|      0
conv3d (Conv3D)     |  (None, 10, 3, 3, 4)|    436|  3888|     36
```
now it outputs
```
Layer (type)        |Output shape         |Param #|Conn #|Neuron #|J/inf (cpu)
--------------------|---------------------|-------|------|--------|-----------
input_1 (InputLayer)|[(None, 10, 3, 3, 4)]|      0|     0|       0|          0
conv3d (Conv3D)     |  (None, 10, 3, 3, 4)|    436| 38880|     360|    0.00034
```

Reply:
> Here's an example that worked correctly with the previous implementation, but doesn't with the current implementation

I don't think there's anything that says that the first implementation was correct and the second is incorrect, it's just a matter of interpretation. There's nothing about that model definition that says the Conv3D operator is operating over time. So the second case assumes that it is _not_ operating over time, and outputs the correct results assuming that is the case. But what we could do is explicitly assume that `Conv1D/3D` are always operating over time (similar to what we do with SpikingActivation). I opted for the more conservative assumption, but I think it'd also be fine to do the latter if we think it'll line up with user's expectations more often (as you say, these are all things that we'll just need to make clear in the documentation one way or the other).

Reply:
You're absolutely right, that it doesn't have to correspond to time. My question was, in the case that it _does_ correspond to time, how do we support that?

I think the only way to Conv layers potentially convolving over time and potentially not is to have the user provide some guidance (e.g. a `timesteps` value) when they create `ModelEnergy`. That's why I went that route. It does mean that it's slightly less automatic, but overall it's more flexible. 

If we are fine losing that flexibility, then I agree that assuming that Conv1D and Conv3D layers are always over time is the best bet. I think that will cover 90% of cases at least (I can't think of any examples that I've seen or done that would use a Conv1D or Conv3D layer not over time).

Reply:
Result of internal discussion: since temporal convolution isn't currently supported in the Nengo frontend or any backends, we will assume that all convolutions are non-temporal. If we add support for temporal convolutions in the future we'd revisit this and add support for it in ModelEnergy.